spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java

```diff
@@ -0,0 +1,192 @@
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    Identifier tableIdent = toIdentifier(args.getString(0), PARAMETERS[0].name());
+
+    return modifyIcebergTable(tableIdent, table -> {
+      RewriteDataFiles action = actions().rewriteDataFiles(table);
+
+      String strategy = args.isNullAt(1) ? null : args.getString(1);
+      SortOrder sortOrder = collectSortOrders(table, args.isNullAt(2) ? null : args.getString(2));
+      action = checkAndApplyStrategy(action, strategy, sortOrder);
+
+      if (!args.isNullAt(3)) {
+        action = checkAndApplyOptions(args, action);
+      }
+
+      String where = args.isNullAt(4) ? null : args.getString(4);
+      action = checkAndApplyFilter(action, where);
+
+      RewriteDataFiles.Result result = action.execute();
+
+      return toOutputRows(result);
+    });
+  }
+
+  private RewriteDataFiles checkAndApplyFilter(RewriteDataFiles action, String where) {
+    if (where != null) {
+      ParserInterface sqlParser = spark().sessionState().sqlParser();
+      try {
+        Expression expression = sqlParser.parseExpression(where);
+        return action.filter(SparkExpressionConverter.convertToIcebergExpression(expression));
+      } catch (ParseException e) {
+        throw new IllegalArgumentException("Cannot parse predicates in where option: " + where);
+      }
+    }
+    return action;
+  }
+
+  private RewriteDataFiles checkAndApplyOptions(InternalRow args, RewriteDataFiles action) {
+    Map<String, String> options = Maps.newHashMap();
+    args.getMap(3).foreach(DataTypes.StringType, DataTypes.StringType,
+            (k, v) -> {
+              options.put(k.toString(), v.toString());
+              return BoxedUnit.UNIT;
+            });
+    return action.options(options);
+  }
+
+  private RewriteDataFiles checkAndApplyStrategy(RewriteDataFiles action, String strategy, SortOrder sortOrder) {
+    if (strategy != null) {
+      if (strategy.equalsIgnoreCase("binpack")) {
+        RewriteDataFiles rewriteDataFiles = action.binPack();
+        if (sortOrder != null) {
+          // calling below method to throw the error as user has set both binpack strategy and sortorder
+          return rewriteDataFiles.sort(sortOrder);
+        }
+        return rewriteDataFiles;
+      } else if (strategy.equalsIgnoreCase("sort")) {
+        if (sortOrder == null) {
+          return action.sort();
+        }
+        return action.sort(sortOrder);
+      } else {
+        throw new IllegalArgumentException("unsupported strategy: " + strategy + ". Only binpack,sort is supported");
+      }
+    } else {
+      // override default binpack to sort, as strategy is not set but sort order is set.
+      if (sortOrder != null) {
+        return action.sort(sortOrder);
+      }
+    }
+    return action;
+  }
+
+  private SortOrder collectSortOrders(Table table, String sortOrderStr) {
+    if (sortOrderStr != null) {
```

Comment:
If we are going to handle the null in this function I would probably just switch this to a 
```
if (sortOrderStr == null) return null; 
```
Rather than wrapping the whole body of the function.

We could also move the null handling to the arg parsing

Reply:
moved to arg parsing